src/librustc_typeck/check/_match.rs

```diff
@@ -167,6 +170,25 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             prior_arm_ty = Some(arm_ty);
         }
 
+        // If all of the arms in the 'match' diverge,
+        // and we're dealing with an actual 'match' block
+        // (as opposed to a 'match' desugared from something else'),
+        // we can emit a better note. Rather than pointing
+        // at a diverging expression in an arbitrary arm,
+        // we can point at the entire 'match' expression
+        match (all_arms_diverge, match_src) {
```

Comment:
Use `if let` here.

Reply:
That requires two `if` blocks - one inside the other - since we also need to check `match_src`. Unfortunately, let-chains aren't implemented yet, so this is the best we can do.

Reply:
How so?
```rust
if let (Diverges::Always { .. }, hir::MatchSource::Normal) = (all_arms_diverge, match_src) {
    ...
}
```
should do the trick.

Reply:
Wow - I had no idea that was possible!